[#147] Allow choosing export type

jtojnar opened the pull request:
As ``hamster-lib`` already supports multiple export backends, this commit only had to add an interface making it accessible.

Currently, CSV, iCalendar and XML are supported by ``hamster-lib``.

The unit tests will have to be added before merging.

Closes: #145 #147

jtojnar committed:
Allow choosing the export type

As ``hamster-lib`` already supports multiple export backends,
this commit only had to add an interface making it accessible.

Currently, CSV, iCalendar and XML are supported.

Closes: #143

jtojnar commented:
![User Interface](https://cloud.githubusercontent.com/assets/705123/20299857/ea8041da-ab1c-11e6-93eb-96f92d1dc719.png)

I had to create a widget for the selector. There is supposed to be [add_choice](https://lazka.github.io/pgi-docs/#Gtk-3.0/interfaces/FileChooser.html#Gtk.FileChooser.add_choice) method but it segfaults for me.

elbenfreund commented:
Thanks for the info. `add_choice` is >3.22 anyway. Seeing that we target >3.10 not using it is a good thing :)
We probably should create an issue though so we may switch to `add_choice` in some later iteration

elbenfreund commented on hamster_gtk/overview/dialogs/export_dialog.py:
```diff
@@ -0,0 +1,102 @@
+
+from gi.repository import Gtk
+
+ExportType = Enum('ExportType', ('CSV', 'ICAL', 'XML'))
```
Why do you use a 'Enum' here instead of just a dict?
Btw: Future iteration should be able to avoid this altogether as the backend will provide a registry for exactly this purpose. [LIB-232](https://projecthamster.atlassian.net/browse/LIB-232)

elbenfreund commented on hamster_gtk/overview/dialogs/export_dialog.py:
```diff
@@ -0,0 +1,102 @@
+
+    def __init__(self, parent):
+        """
+        Initialize headerbar.
```
'headerbar' should be 'dialog'

elbenfreund commented on hamster_gtk/overview/dialogs/export_dialog.py:
```diff
@@ -0,0 +1,102 @@
+            Gtk.FileChooserAction.SAVE, (Gtk.STOCK_CANCEL, Gtk.ResponseType.CANCEL,
+                                         Gtk.STOCK_SAVE, Gtk.ResponseType.OK))
+
+        self._export_type_chooser = Gtk.ComboBoxText()
```
It would be better to wrap all the `type_chooser` related code into a private method like `self._get_type_choose` that returns the deals with creating label and dropdown and just returns the grid.

elbenfreund commented on hamster_gtk/overview/dialogs/export_dialog.py:
```diff
@@ -0,0 +1,102 @@
+                                         Gtk.STOCK_SAVE, Gtk.ResponseType.OK))
+
+        self._export_type_chooser = Gtk.ComboBoxText()
+        self._export_type_chooser.append(str(ExportType.CSV), _("CSV"))
```
Export types do not need localization or?

jtojnar replied:
I suppose some locales might want to translate it.

elbenfreund commented on hamster_gtk/overview/dialogs/export_dialog.py:
```diff
@@ -0,0 +1,102 @@
+        self._export_type_chooser.append(str(ExportType.CSV), _("CSV"))
+        self._export_type_chooser.append(str(ExportType.ICAL), _("iCal"))
+        self._export_type_chooser.append(str(ExportType.XML), _("XML"))
+        self._export_type_chooser.connect('changed', self._export_type_changed)
```
callback should be called `_on_export_type_changed`.

elbenfreund commented on hamster_gtk/overview/dialogs/export_dialog.py:
```diff
@@ -0,0 +1,102 @@
+        self._export_type_chooser.append(str(ExportType.XML), _("XML"))
+        self._export_type_chooser.connect('changed', self._export_type_changed)
+
+        export_type_label = Gtk.Label(_("Export as file _type:"))
```
I would prefer `Export as`. I think 'file type' is misleading. They are all 'text files' its just semantics of its content that are different. Alternatively I could live with 'Export type...'.

elbenfreund commented on hamster_gtk/overview/dialogs/export_dialog.py:
```diff
@@ -0,0 +1,102 @@
+        export_options.attach(export_type_label, 0, 0, 1, 1)
+        export_options.attach_next_to(self._export_type_chooser, export_type_label,
+                                      Gtk.PositionType.RIGHT, 1, 1)
+        export_options.show_all()
```
Wouldn't it be better to call `self.show_all`? This should traverse down to the combobox as well.

jtojnar replied:
There might be some intentionally hidden widgets in the dialog. I would rather show the narrowest set of widgets possible.

elbenfreund commented on hamster_gtk/overview/dialogs/export_dialog.py:
```diff
@@ -0,0 +1,102 @@
+        self.set_current_name(_("hamster_export"))
+        self._export_type_chooser.set_active_id(str(ExportType.CSV))
+
+    def _export_type_changed(self, combobox):
```
should be named _on_export_type_changed

elbenfreund commented on hamster_gtk/overview/dialogs/__init__.py:
```diff
@@ -0,0 +1,21 @@
+
+"""This module provides an overview dialog allowing access to all facts, and an export dialog."""
+
+from .export_dialog import ExportDialog, export_writers  # NOQA
```
Do we really want to make ``export_writers`` public?
wouldn't it suffice to put it somewhere around [here](https://github.com/projecthamster/hamster-gtk/pull/144/files#diff-1e116303a380135b42893c60abe25a72R255)?

elbenfreund commented on hamster_gtk/overview/dialogs/export_dialog.py:
```diff
@@ -0,0 +1,84 @@
+
+from hamster_gtk.preferences.widgets import PreferencesGrid
+
+export_writers = {'csv': reports.TSVWriter, 'ical': reports.ICALWriter, 'xml': reports.XMLWriter}
```
``csv`` should be ``tsv``

jtojnar replied:
Should the extension also be `tsv`? It was `csv` previously: https://github.com/projecthamster/hamster-gtk/pull/144/files#diff-3d1746373db586ae8ad2273cc902a5cbL114

elbenfreund replied:
yes please.

elbenfreund commented on hamster_gtk/overview/dialogs/export_dialog.py:
```diff
@@ -0,0 +1,84 @@
+        self._export_format_chooser = self._get_export_format_chooser()
+        self._export_format_chooser.connect('changed', self._on_export_format_changed)
+
+        export_options = PreferencesGrid(
```
This is really messy. Either ``PreferencesGrid`` needs to be renamed/relocated to reflect its more general usage or it should not be used here.

elbenfreund commented on hamster_gtk/overview/widgets/misc.py:
```diff
@@ -108,14 +109,12 @@ def _on_export_button_clicked(self, button):
             target_path = dialog.get_filename()
-            parent._export_facts(target_path)
+            target_format = dialog.get_export_format()
+            parent._export_facts(target_format, target_path)
```
L116 and L117 could probably be merged into one...

jtojnar replied:
Should that also include line 115?

elbenfreund replied:
yes

elbenfreund commented on tests/overview/test_dialogs.py:
```diff
@@ -38,9 +39,9 @@ def test__export_facts(self, overview_dialog, tmpdir, mocker):
+        writer = mocker.patch('hamster_gtk.overview.dialogs.overview_dialog.reports.TSVWriter')
         overview_dialog._get_facts = mocker.MagicMock(return_value={})
-        result = overview_dialog._export_facts(tmpdir.strpath)
+        result = overview_dialog._export_facts('csv', tmpdir.strpath)
```
this test should probably be parametrized over all supported 'export strings'.

jtojnar committed:
Rename PreferencesGrid to LabeledWidgetsGrid

As we make use of the custom Grid outside the preferences context the
class was renamed to reflect its more general nature. As a consequence
the existing code had to be adjusted.
